nltk/translate/chrf_score.py

```diff
@@ -110,28 +132,59 @@ def corpus_chrf(list_of_references, hypotheses, min_len=1, max_len=6, beta=3.0):
     :type max_len: int
     :param beta: the parameter to assign more importance to recall over precision
     :type beta: float
+    :param ignore_whitespace: ignore whitespace characters in scoring
+    :type ignore_whitespace: bool
     :return: the sentence level CHRF score.
     :rtype: float
     """
 
-    assert len(list_of_references) == len(hypotheses), "The number of hypotheses and their references should be the same"
+    assert len(references) == len(hypotheses), (
+        "The number of hypotheses and their references should be the same")
+
+    # Keep f-scores for each n-gram order separate
+    n_gram_fscores = {k: [] for k in range(min_len, max_len + 1)}
+
+    num_sents = 0  # we don't call len() in case these are generators
 
     # Iterate through each hypothesis and their corresponding references.
-    for reference, hypothesis in zip(list_of_references, hypotheses):
-        # Cheating condition to allow users to input strings instead of tokens.
-        if type(reference) and type(hypothesis) != str:
-            reference, hypothesis = ' '.join(reference), ' '.join(hypothesis)
-        # For each order of ngram, calculate the no. of ngram matches and
-        # keep track of no. of ngram in references.
-        ref_ngrams = Counter(everygrams(reference, min_len, max_len))
-        hyp_ngrams = Counter(everygrams(hypothesis, min_len, max_len))
-        overlap_ngrams = ref_ngrams & hyp_ngrams
-        tp = sum(overlap_ngrams.values()) # True positives.
-        tpfp = sum(hyp_ngrams.values()) # True positives + False positives.
-        tffn = sum(ref_ngrams.values()) # True posities + False negatives.
-
-    precision = tp / tpfp
-    recall = tp / tffn
-    factor = beta**2
-    score = (1+ factor ) * (precision * recall) / ( factor * precision + recall)
-    return score
+    for reference, hypothesis in zip(references, hypotheses):
+        num_sents += 1
+
+        # preprocess both reference and hypothesis
+        reference = _preprocess(reference, ignore_whitespace)
+        hypothesis = _preprocess(hypothesis, ignore_whitespace)
```

Comment:
I still find this weird. I think it's inconsistent with the rest of the scores in NLTK where the default is to assume that the inputs are preprocess and the scoring functions should focus on the numerical computation as much as possible.

Reply:
Maybe it's time to really refactor the `nltk.translate.*_score.py` to a proper module that handles the preprocessing then the computation.

Reply:
Isn't that out of the scope of this PR?

Reply:
Hmmm, let sees what we can do. I'll relook at the changes and then refactor the code again. It's easier to do it right for each metric and keeping in mind that someday we'll have this unified across all metrics.